### ISSUE TITLE
Install developer tools to $XDG_DATA_HOME as default.

### ISSUE THREAD

**Issue by marcinant** [ubuntu/ubuntu-make#81]
I think that it could be nice to keep home directory clean and install developer tools to ~/.local/share folder. Binaries could go to ~/.local/bin.
This path could be also added to $PATH with this code in ~/.profile
```
if [ -d "$HOME/.local/bin" ] ; then
    PATH="$HOME/.local/bin:$PATH"
fi
```

In this way applications could be available to run from command line.

**Comment by didrocks** [comment]
Hi marcinant,

Thanks for the suggestion! I guess installing applications in other places than ~/tools makes sense.
Some things to note though:
- XDG_DATA_HOME isn't supposed to contain softwares, but user's data. From the xdg spec (http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html): "There is a single base directory relative to which user-specific data files should be written. This directory is defined by the environment variable $XDG_DATA_HOME."
- I don't want to separate binaries from the software package (in an eventual .local/bin).  That breaks most of the auto-update system by the usptream softwares.
There is a tool to add to user's $PATH on a per framework basics in ubuntu-make. If you feel that some are missing, we can work on this (we don't add them if there is a .desktop file pointing to that binary usually).

If we can find a better default path for installing softwares (and compatible with the XDG specs) though, I'm opened to the data. Any suggestions?

**Comment by marcinant** [comment]
Well, if you want to be so attached to the standards and recommendations then in fact `umake` should not install any software in userspace. Everything should go to: /opt or /usr/local. However we all know it's not the point as `umake` is for an easy rootless installation and **updates**.

[FHS](http://www.pathname.com/fhs/pub/fhs-2.3.html#HOMEUSERHOMEDIRECTORIES) says:
>/home is a fairly standard concept, but it is clearly a site-specific filesystem. [9] The setup will differ >from host to host. **Therefore, no program should rely on this location.**

There are documents such as [Ubuntu HomeFolder](https://help.ubuntu.com/community/HomeFolder#Installing_Software_Into_The_Home_Directory)
or [Arch filesystem hierarchy](https://wiki.archlinux.org/index.php/Arch_filesystem_hierarchy#.2Fhome:_User_home_directories):
> These are the user home directories, which can be found under '/home/$USER' (~/). Within their home directory, a user can write files, delete them, **install programs**, etc.

However there is also [this](http://standards.freedesktop.org/menu-spec/latest/apcs02.html) document which says:

>If an application is intended to be installed by an unprivileged user for exclusive use by that user >only then $XDG_DATA_HOME should be used as value for datadir and $XDG_CONFIG_HOME >should be used as value for sysconfdir. If $XDG_DATA_HOME is not set, the default value of >$HOME/.local/share should be used for it. If $XDG_CONFIG_HOME is not set, the default value of >$HOME/.config should be used for it. 

This is why I recommended to change ~/tools to $XDG_DATA_HOME

**Comment by Tinche** [comment]
I've been thinking of the proper way to deal with binaries we want added to a user's path since using golang support.

Honestly I think trying to dynamically add a separate directory for each tool to the path is too fragile. The .profile thing works only for bash (certainly not for fish), and Ubuntu Make is intended for developers, so who the hell knows which shell the users will be using. Also I fear dynamically editing files from Python may have corner cases.

The default Ubuntu .profile (/etc/skel/.profile) adds $HOME/bin to the users path if it exists. Maybe we should leverage this? Just symlink binaries from their tool directories to $HOME/bin. It would be nice if we could dynamically check at installation time if $HOME/bin is on the user's path, and if not just *ask* the user to add it manually.

I like ~/tools for other tool files.

**Comment by marcinant** [comment]
> The default Ubuntu .profile (/etc/skel/.profile) adds $HOME/bin to the users path if it exists. 

I mentioned the document describing this behavior in my comment: [Ubuntu HomeFolder](https://help.ubuntu.com/community/HomeFolder#Installing_Software_Into_The_Home_Directory)

However Freedesktop recommendation is to put applications to $XDG_DATA_HOME.

Personally i think that $XDG_DATA_HOME is ok and icon launchers should point to applications there.

~/bin is a kind of additional feature which allows to launch applications from command line.

Unfortunately binaries should not go to ~/bin directly as they will conflict with system wide apps.
For example Firefox Developer Edition is going to conflict with Firefox installed from *.deb.

**Comment by didrocks** [comment]
On the "I don't like" -> I guess nothing prevents you to use that, we are really talking about default, what most developer would use.

I like Tinche's idea's of ~/bin. I'm using it myself for random home-made tools. The idea is that you have your binaries, with ubuntu make installed and that's the default, so it can works (having symlinks to the real bins in ~/tools). The firefox dev edition is a special case, and we can workaround it in naming the symlink in ~/bin/firefox-dev -> ~/tools/web/firefox…

We need then to develop in the framework a facility to provide a list of binaries to symlinks in ~/bin (and some optional way to have them eventually renamed while symlinking). That way, we don't lose the update functionality. Tin, do you think you have the time to prototype this?

Then, the question of ~/tools vs $XDG_DATA_HOME for default is really a question of taste it seems…

**Comment by marcinant** [comment]
On the "I don't like"... sure :) @Tinche is ok with ~/tools while I'm not. So it's 1:1 (50%:50%) ;)
I just customized install path to put apps to $XDG_DATA_HOME.

However Freedesktop recommends as I mentioned before:
>If an application is intended to be installed by an unprivileged user for exclusive use by that user only then $XDG_DATA_HOME should be used as value for datadir and $XDG_CONFIG_HOME should be used as value for sysconfdir.

So, it is not only a matter of taste.

Ubuntu recommends binaries in ~/bin and Freedesktop recommends /usr/share in $XDG_DATA_HOME. Personally I think that installing applications with umake to $XDG_DATA_HOME with custom symlinks pointing to ~/bin is the most elegant solution.

**Comment by Tinche** [comment]
@didrocks Ugh I'd love to but I already promised someone I would try working on something else. So little time...

**Comment by didrocks** [comment]
ok, I think this whole makes sense with the additional quotation from Martin. However, my only remaining objection to this plan for now is if some IDEs or any frameworks used ~/.local/share/<appname> to store actual application user data, can't we get any conflict on this? What's your suggestion?

Then, symlinks with ~/bin sounds indeed elegant.

 can deal with it, apart if martin wants to give a try? I'm happy to assist

**Comment by marcinant** [comment]
> ok, I think this whole makes sense with the additional quotation from Martin. However, my only remaining objection to this plan for now is if some IDEs or any frameworks used ~/.local/share/ to store actual application user data, can't we get any conflict on this? What's your suggestion?

$XDG_DATA_HOME is base directory. Tools installed by umake can go some subdir there.
Personally I keep them in... tools dir. ~/.local/share/tools
So, it's just two levels deep and nothing more.

It can be also something like:
```
~/.local/share/umake/android
~/.local/share/umake/ide
~/.local/share/umake/web
```
This ~/bin thing is ok as you don't need to modify ~/.profile. On Ubuntu it just works.

**Comment by didrocks** [comment]
ok, so you see umake as the "container" (like steam or wine) for holding the data part. Sounds good enough to me.

Do you have the skill/want to work on it? I'm happy to provide tutoring (in particular on adapating the tests to it).

**Comment by oijazsh** [comment]
I can work on this, provided the choice of $XDG_DATA_HOME/umake as default is agreeable for most.

**Comment by didrocks** [comment]
This is now released since Ubuntu Make 15.11. Closing the bug now.
